fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PushDownCountThroughJoinOneSide.java

```diff
@@ -0,0 +1,216 @@
+            }
+        }
+
+        Set<Slot> leftGroupBy = new HashSet<>();
+        Set<Slot> rightGroupBy = new HashSet<>();
+        for (Expression e : agg.getGroupByExpressions()) {
+            Slot slot = (Slot) e;
+            if (leftOutput.contains(slot)) {
+                leftGroupBy.add(slot);
+            } else if (rightOutput.contains(slot)) {
+                rightGroupBy.add(slot);
+            } else {
+                return null;
+            }
+        }
+        join.getHashJoinConjuncts().forEach(e -> e.getInputSlots().forEach(slot -> {
+            if (leftOutput.contains(slot)) {
+                leftGroupBy.add(slot);
+            } else if (rightOutput.contains(slot)) {
+                rightGroupBy.add(slot);
+            } else {
+                throw new IllegalStateException("Slot " + slot + " not found in join output");
+            }
+        }));
+
+        Plan left = join.left();
+        Plan right = join.right();
+
+        Map<Slot, NamedExpression> leftCntSlotToOutput = new HashMap<>();
+        Map<Slot, NamedExpression> rightCntSlotToOutput = new HashMap<>();
+
+        // left Count agg
+        if (!leftCounts.isEmpty()) {
+            Builder<NamedExpression> leftCntAggOutputBuilder = ImmutableList.<NamedExpression>builder()
+                    .addAll(leftGroupBy);
+            leftCounts.forEach(func -> {
+                Alias alias = func.alias(func.getName());
+                leftCntSlotToOutput.put((Slot) func.child(0), alias);
+                leftCntAggOutputBuilder.add(alias);
+            });
+            left = new LogicalAggregate<>(ImmutableList.copyOf(leftGroupBy), leftCntAggOutputBuilder.build(),
+                    join.left());
+        }
+
+        // right Count agg
+        if (!rightCounts.isEmpty()) {
+            Builder<NamedExpression> rightCntAggOutputBuilder = ImmutableList.<NamedExpression>builder()
+                    .addAll(rightGroupBy);
+            rightCounts.forEach(func -> {
+                Alias alias = func.alias(func.getName());
+                rightCntSlotToOutput.put((Slot) func.child(0), alias);
+                rightCntAggOutputBuilder.add(alias);
+            });
+
+            right = new LogicalAggregate<>(ImmutableList.copyOf(rightGroupBy), rightCntAggOutputBuilder.build(),
+                    join.right());
+        }
+
+        Preconditions.checkState(left != join.left() || right != join.right());
+        Plan newJoin = join.withChildren(left, right);
+
+        // top Sum agg
+        // count(slot) -> sum( count(slot) as cnt )
+        List<NamedExpression> newOutputExprs = new ArrayList<>();
+        for (NamedExpression ne : agg.getOutputExpressions()) {
+            if (ne instanceof Alias && ((Alias) ne).child() instanceof Count) {
+                Count oldTopCnt = (Count) ((Alias) ne).child();
+
+                Slot slot = (Slot) oldTopCnt.child(0);
+                if (leftCntSlotToOutput.containsKey(slot)) {
```

Comment:
would we do some pre-processing ahead, such as judge which side will push down, and at this point, we just do the transformation without the additional 'if-else'?